tools/fp20compiler/main.cpp

```diff
@@ -34,27 +34,82 @@ nvparse_errors errors;
 int line_number;
 char * myin = 0;
 
+static char* find_at_line_start(char* haystack, char* cursor, const char* needle) {
+    cursor = strstr(cursor, needle);
+    while (cursor != NULL) {
+
+        // Accept if candidate is at beginning of file or line
+        if ((cursor == haystack) || (cursor[-1] == '\n')) {
+            return cursor;
+        }
+
+        cursor = strstr(cursor + 1, needle);
+    }
+    return NULL;
+}
+
 void translate(const char* s) {
     char* ns = strdup(s);
 
-    // printf("-- %s\b", s);
+    // Keep a cursor for line-counting
+    char* line_cursor = ns;
+    unsigned int line = 1;
 
-    char* ts_start = strstr(ns, "!!TS1.0");
-    if (ts_start) {
-        char* ts_end = strstr(ts_start, "// End of program");
-        assert(ts_end);
-        *ts_end = 0;
+    // Look for the first shader magic
+    char* shader_start = find_at_line_start(ns, ns, "!!");
 
-        ts10_init(ts_start);
-        ts10_parse();
-
-        *ts_end = '/';
+    // Warn the user if we couldn't find any shader at all
+    if (shader_start == NULL) {
+        fprintf(stderr, "no shaders found\n");
     }
 
-    char* rc_start = strstr(ns, "!!RC1.0");
-    if (rc_start) {
-        rc10_init(rc_start);
-        rc10_parse();
+    // Loop until we can't find a shader anymore
+    while (shader_start != NULL) {
+
+        // The next shader magic marks the end of the current shader; find it
+        char* line_end = strchr(shader_start, '\n');
+        char* next_shader_start = find_at_line_start(ns, line_end + 1, "!!");
```

Comment:
`line_end` could be `NULL` at this point. `next_shader_start` must be `NULL` then.
A check has to be added.

Alternatively, we could report an error for this case.
However, I'd like to avoid that for now (as nvparse might also report this as an error - I'm not sure yet).

Reply:
I've fixed this and tested with `echo -n '!!RC1.0' > test.ps.asm`